### ISSUE TITLE
AnyIO implementation of asynchronous streams

### ISSUE THREAD

**Issue by thedrow** [python-streamz/streamz#320]
Using streamz with an event loop such as [`trio`](https://trio.readthedocs.io/en/stable/) is currently not possible since Tornado only supports AsyncIO if it is available.

If we had an implementation that uses [`AnyIO`](https://anyio.readthedocs.io/), we could use any event loop we'd desire including tornado.

**Comment by CJ-Wright** [comment]
What would that entail?

**Comment by CJ-Wright** [comment]
Also how would that interact with dask?

**Comment by martindurant** [comment]
Indeed, a dask client requires an event loop. It is possible to supply an event loop instance at client creation, and I don't know exactly what it required of that loop instance - probably anything asyncio compliant. However, dask's loop can also run in a separate thread, so maybe it is possible to have two loops going (which may get complicated when actually dealing with dask-streamz).

**Comment by thedrow** [comment]
It's possible but not necessary if you aren't using dask.

**Comment by martindurant** [comment]
Have you tried using the anyIO (or other) loop for dask too? That would be the favourite solution if it works. Otherwise, I think I do basically agree that streamz should work regardless, but in practice, it may be tricky to demarcate dask-free parts of the code.

**Comment by thedrow** [comment]
There is no AnyIO implementation for dask.
There's an AsyncIO implementation and a Tornado implementation.
I have never used dask so no on both accounts.

It might be possible to run dask with [`trio-asyncio`](https://trio-asyncio.readthedocs.io/en/latest/) but I haven't really checked.

I want to implement streamz on another framework.

**Comment by thedrow** [comment]
This framework should not tie itself to a specific event loop.
What if the next best thing comes around and everybody wants to use it?

**Comment by martindurant** [comment]
True, but being able to use Dask is important too. If you can implement anyIO (or something else) for the non-Dask code without greatly complicating the code, we would consider it. Of course we would still prefer smooth interaction between streamz and Dask on any io-loop.

**Comment by thedrow** [comment]
If you'd like to integrate Dask with other event loops I suggest taking it upstream.

I took a look at the code. It is currently impossible since you are tightly coupled with the Tornado event loop.

**Comment by martindurant** [comment]
Would uvloop, which is asyncio compatible, be a solution for you?

**Comment by dhirschfeld** [comment]
***xref:*** https://github.com/python-trio/trio-asyncio/issues/22#issuecomment-586974857

**Comment by thedrow** [comment]
> Would uvloop, which is asyncio compatible, be a solution for you?

No. I need trio support.

**Comment by Crypto7816** [comment]
> Would uvloop, which is asyncio compatible, be a solution for you?

Could you please give some example of using python default asynicio and uvloop?

**Comment by martindurant** [comment]
https://streamz.readthedocs.io/en/latest/async.html has asyncio examples and tornado syntax too.

To use a uvloop, you can make it the default one with something like
```python
asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
```
and all further use can be the same. I don't know if there's any further configuration or control you can do with it.
